### PR TITLE
llama : first attempt to implement vision API (WIP)

### LINKED ISSUE #8010: server: Bring back multimodal support
Multimodal has been removed since https://github.com/ggerganov/llama.cpp/pull/5882

## Current llama.cpp multimodal roadmap

(update 9th april 2025)

- `mtmd` (**M**ul**T**i-**M**o**D**al) library (top prio 🔥 )
    - [x] Implement `libmtmd`: https://github.com/ggml-org/llama.cpp/pull/12849
    - [x] Support more models via `libmtmd` (top prio 🔥 ) : https://github.com/ggml-org/llama.cpp/pull/13012
    - [x] Support M-RoPE models via `libmtmd` (Qwen2VL, Qwen2.5VL) : https://github.com/ggml-org/llama.cpp/pull/13141
    - [ ] Support audio input
    - [x] Use smart pointer in `clip.cpp` to avoid mem leak: https://github.com/ggml-org/llama.cpp/pull/12869
    - [x] ~~Add wrapper for `stb_image` to avoid polluting project with the big header file~~ --> Probably don't need since we're already having some helper in `libmtmd` acting as wrapper for stb_image
    - [x] Unify conversion scripts --> best case scenario: having `convert_hf_to_gguf.py` that can output both text + vision GGUF files --> introduced in https://github.com/ggml-org/llama.cpp/pull/13023
    - [x] Remove BOI / EOI token embeddings from clip.cpp (used by glm-edge): https://github.com/ggml-org/llama.cpp/pull/13081
    - [x] Refactor documentations (find a way to reduce number of README files): https://github.com/ggml-org/llama.cpp/pull/13055
- Implement `libmtmd` in server API and server web UI (top prio 🔥 )
   - [x] Publish first proposal: https://github.com/ggml-org/llama.cpp/pull/12898
   - [x] User can upload image from UI ( + drag-and-drop)
   - [x] Nice-to-have: Better KV caching strategy (TBD)
   - [x] Nice-to-have: allow loading remote image (may come with security risk)
   - [ ] Update the [security policy](https://github.com/ggml-org/llama.cpp/security/policy), make it clear that bugs related to 3rd party lib (like `stb_image`) should be reported to upstream, not in llama.cpp
- [x] Unify all vision CLI (like `minicpmv-cli`, `gemma3-cli`, etc) into a single CLI
- [ ] Add deprecation notice for `llava.h` (we will remove libllava) and `clip.h` (clip is now internal-only)
- [ ] Experimental support for audio input: https://github.com/ggml-org/llama.cpp/pull/12745
- [ ] (far in the future) implement `llama_multimodal` API that supports image, audio and more!

### PR DESCRIPTION
**Pull request by ngxson**
(Hopefully) fix #8010

> [!IMPORTANT]
> This is still WIP, only `simple` example is working
> Collaborators are encouraged to discuss and give feedback on this

## Motivation

Currently, the vision capability is provided by `llava` example, which is a CLIP implementation in ggml. While it's a good start, the API need some refactoring to be cleaner and more future-proof.

Inspired by current rework on sampling API, I propose to move the CLIP implementation into the main `libllama`, providing user a stable, easy-to-use API like what we did for `llama_encode`

The goals of this refactoring are:
- Provide a good API and code architecture for more models to come in the future
- Single gguf file for both vision+language (so, no more model surgery)
- Only llava for now (because it simple for me to understand)
- Have `llama-cli` accept image input

The no-goals:
- No change to `llama-server`. It will be another PR
- No minicpm, llama-3.2-vision, phi-3-vision, etc. Again, will be another PR

## Plan

- [x] define the plan:
  - gguf metadata and tensor naming scheme
  - define API to be exposed from `libllama`
- [x] upgrade `convert_hf_to_gguf.py` to support llava --> not an ideal implementation, but kinda works
- [ ] extend `llama_model` and `llama_context` to hold vision-related data
- [ ] add `llama-vision.{cpp|h}`
- [ ] add image capability to `llama-cli`

### Naming scheme

For metadata, we will add `vision.*` namespace.
- `vision.type`: the type of vision encoder. We only support `"clip"` for now (not sure if there are any other implementation out there)
- `vision.*`: other params for vision encoding, for example patch size, image size, etc
- `vision.clip.*`: CLIP-related params

Example:

```
vision.type = 'clip'
vision.image_size = 336
vision.patch_size = 14
vision.clip.architecture = 'llava'
vision.clip.block_count = 24
vision.clip.embedding_length = 1024
vision.clip.feed_forward_length = 4096
vision.clip.attention.head_count = 16
```

For tensor naming scheme, we will prefix all vision-related tensor with `v.enc.*`. For example:

```
v.mmproj_a.bias
v.mmproj_a.weight
v.enc.embd.cls
v.enc.embd.patch.weight
v.enc.embd.pos.weight
v.enc.blk.0.input_norm.bias
v.enc.blk.0.input_norm.weight
```

### API

`libllama`will be responsible for:
- Accepting bitmap image (RGB format) and split it into patches
- It will NOT process specific format like PNG, JPG, etc. User must convert these format into bitmap (for example, using STB) before giving to llama
- The API returns embeddings that can be add to a language batch

```cpp
// represent an RGB image
// size of data must be equal to 3*nx*ny
struct llama_img {
    uint32_t nx;
    uint32_t ny;
    unsigned char * data;
};

typedef struct llama_img_batch {
    int32_t     n_imgs;
    llama_img * imgs;
    // add other things in future?
}

// encode image into embeddings
int32_t llama_vision_encode(struct llama_context * ctx, llama_img_batch * batch);

// get output embeddings, to be put into language batch
float * llama_vision_get_embeddings(struct llama_context * ctx, int32_t idx);

```

---

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] High

### CODE REVIEW

File: gguf-py/gguf/constants.py [hunk 1]
```diff
@@ -178,6 +178,28 @@ class Adapter:
         TYPE       = "adapter.type"
         LORA_ALPHA = "adapter.lora.alpha"
 
+    class Vision:
+        # only support vision.type = "clip" for now
```
Comment (ashvinnihalani) [inline gguf-py/gguf/constants.py:182]: Probably better to be very specific here you are supporting ViT. HuggingFace has also moved away from calling every generic terms. Also I don't think the purpose is to support actual CLIP inference.

Comment (ngxson) [reply gguf-py/gguf/constants.py:182]: Yeah, agree. I think for now it's safer to call this `clip-vit` to reflect the base implementation is `openai/clip-vit-*`

Atm it's quite complicated for me to drop the `clip_` prefix in all functions. But hey at least the file name is now `llama-vision.{cpp|h}`, which should reflect that we can support ViT and also other things to come in the future.

File: examples/simple/simple.cpp [hunk 2]
```diff
@@ -84,41 +87,93 @@ int main(int argc, char ** argv) {
...
+        n_cur = 0;
+        llama_batch_clear(batch);
+        for (auto t : t1) { llama_batch_add(batch, t, n_cur, { 0 }, false); n_cur++; }
+        llama_decode(ctx, batch);
+
+        n_cur = t1.size() + 576;
+        llama_batch_clear(batch);
+        printf("pos %d\n", n_cur);
+        for (auto t : t2) { llama_batch_add(batch, t, n_cur, { 0 }, false); n_cur++; }
+        batch.logits[batch.n_tokens - 1] = true;
```
Comment (ngxson) [inline examples/simple/simple.cpp:156]: @ggerganov I observed something weird here that I can't understand, so I would like to ask if you know why.

What I'm trying to do here is to test if I can call `llama_decode` on different parts of the sentence out-of-order. That means instead of evaluating `text,image,text`, I try evaluating `image,text,text`. In theory, this should work because the position is set correctly for all input tokens/embeddings, just the placement in KV cache is different.

But in reality, `llama_decode` on a batch containing `embd` doesn't work correctly if the **KV cache is empty**. In this example, I need to evaluate at least one token for it to work.

I'd be happy to provide more info to help pin-point the problem. Thank you.

Comment (ggerganov) [reply examples/simple/simple.cpp:156]: > But in reality, llama_decode on a batch containing embd doesn't work correctly if the KV cache is empty.

What error do you get?

Comment (ngxson) [reply examples/simple/simple.cpp:156]: There is no error, but it the generated text is quite irrelevant.
- Expected output: `The image features a beautiful view of the Eiffel Tower`
- Without the first single-token `llama_decode`: `The image features a cityscape with a tall building prominently visible in the center. The building appears to be lit up, possibly at night or during` --> clearly wrong because I provided a daytime image

Please note that if the code block follow order of `text,image,text`, then it will work correctly (generate expected output).

I have a feeling that the KV for embedding is truncated somehow, but can't find a way to very this. I tried playing with `llama_kv_cache_view` but it seems like it's all OK.

<details>
  <summary>show code</summary>

```cpp
        llama_kv_cache_view v = llama_kv_cache_view_init(ctx, 1);
        llama_kv_cache_view_update(ctx, &v);
        printf("n_cells %d\n", v.n_cells);
        for (int i = 0; i < 16; i++) {
            printf("cell %d --> pos %d (seq %d)\n", i, v.cells[i].pos, v.cells_sequences[0]);
        }
        for (int i = 570; i < 570+16; i++) {
            printf("cell %d --> pos %d (seq %d)\n", i, v.cells[i].pos, v.cells_sequences[0]);
        }
        for (int i = n_cur-8; i < n_cur+8; i++) {
            printf("cell %d --> pos %d (seq %d)\n", i, v.cells[i].pos, v.cells_sequences[0]);
        }
```

</details>


(I'm uploading the model so that you can test it. Will drop a link here when it's available)

Comment (ngxson) [reply examples/simple/simple.cpp:156]: Link to gguf: https://huggingface.co/ngxson/test-llava-will-be-deleted-soon/blob/main/model.gguf

Command: `./llama-simple -m ../models/llava-1.5-7b-hf/model.gguf -c 1024 -v`

Comment (ggerganov) [reply examples/simple/simple.cpp:156]: I don't think that `text,image,text` is equivalent to `image,text,text` regardless if the positions are correct. But before we discuss that, it seems to me that the positions are off by one due to the BOS token:

```c++
        n_cur = 0;
        llama_batch_clear(batch);
        llama_batch_add(batch, 1, 0, { 0 }, false);
        llama_decode(ctx, batch);

        n_cur = t1.size();
        llama_batch_clear(batch);
        llama_batch batch0 = {int32_t(576), nullptr, _test_get_img_embd(ctx), nullptr, nullptr, nullptr, nullptr, 1 + n_cur, 1, 0, };
        llama_decode(ctx, batch0);

        n_cur = 0;
        llama_batch_clear(batch);
        for (auto t : t1) { llama_batch_add(batch, t, 1+ n_cur, { 0 }, false); n_cur++; }
        llama_decode(ctx, batch);

        n_cur = t1.size() + 576;
        llama_batch_clear(batch);
        printf("pos %d\n", n_cur);
        for (auto t : t2) { llama_batch_add(batch, t, 1+ n_cur, { 0 }, false); n_cur++; }
        batch.logits[batch.n_tokens - 1] = true;
```

Comment (ngxson) [reply examples/simple/simple.cpp:156]: The BOS is added manually, `t1.insert(t1.begin(), 1);`

The `text,image,text` sequence looks like this:

```cpp
        auto t1 = ::llama_tokenize(ctx, "A chat between a curious human and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the human's questions.\nUSER:", false);
        auto t2 = ::llama_tokenize(ctx, "\nwhat did you see?\nASSISTANT:", false);
        t1.insert(t1.begin(), 1);

        n_cur = 0;
        llama_batch_clear(batch);
        for (auto t : t1) { llama_batch_add(batch, t, n_cur, { 0 }, false); n_cur++; }
        llama_decode(ctx, batch);

        n_cur = t1.size();
        llama_batch_clear(batch);
        llama_batch batch0 = {int32_t(576), nullptr, _test_get_img_embd(ctx), nullptr, nullptr, nullptr, nullptr, n_cur, 1, 0, };
        llama_decode(ctx, batch0);

        n_cur = t1.size() + 576;
        llama_batch_clear(batch);
        printf("pos %d\n", n_cur);
        for (auto t : t2) { llama_batch_add(batch, t, n_cur, { 0 }, false); n_cur++; }
        batch.logits[batch.n_tokens - 1] = true;
```

By simply moving the code block `llama_decode(ctx, batch0);` to the beginning, you can observe the different (even though all token positions stays the same)

Comment (ngxson) [reply examples/simple/simple.cpp:156]: > But before we discuss that, it seems to me that the positions are off by one due to the BOS token

Just to clarify a bit, the single-token llama_decode is there just for testing. Indeed, adding that will effective make the sequence to have 2 BOS at the same position 0. But for some reasons, the output is the same with single BOS: `bos,text,image,text`

Comment (ngxson) [reply examples/simple/simple.cpp:156]: Thanks for your response. FYI, I tried printing the `inp_KQ_mask` and turns out it's the problem. In the second case `image, text0, text1`, somehow the `text1` can't attend to tokens in `text1`.

What's weird is that, if now I have 4 batches: `bos,image,text0,text1`, then it will work correctly (`bos` is the batch that contains one single BOS token)

I also set the `n_batch` and `n_ubatch` to 1024 to make sure the image fit into one batch, but it doesn't solve my issue.

> One thing to be careful is when computing the `image` batch, what mask should we use?

In the case of llava model, let's imagine that the image batch is just a batch containing 576 tokens, so let's call it `textI` instead of `image`. So my assumption is `text0,textI,text1` should be equivalent to `textI,text0,text1` (?). When generating the next token, we simply add all tokens that we already evaluated to the mask, so it should be a auto-aggressive mask.

Perhaps PaliGemma is different because it requires non-causal.

---

I also tried setting `llama_set_causal_attn` to false before evaluating `image` batch, but that doesn't help either.

Comment (ngxson) [reply examples/simple/simple.cpp:156]: For now, I think a hacky way would be to internally populate image tokens into the batch in `llama_decode`. I'm also thinking about possibility of extending `build_inp_pos` to accept both token and embedding input at the same time. What do you think? (also cc @abetlen since that what your #7553 is about)

File: src/llama-vision.cpp [hunk 3]
```diff
@@ -0,0 +1,867 @@
...
+        for (int x = 0; x < new_width; ++x) {
+            for (int c = 0; c < 3; ++c) {
+                padded_image.buf[3 * ((y + pad_y) * target_width + (x + pad_x)) + c] = resized_image.buf[3 * (y * new_width + x) + c];
+            }
+        }
+    }
+    image_output = std::move(padded_image);
+}
+
+static void normalize_image_u8_to_f32(const clip_image_u8 src, clip_image_f32 dst, const std::array<float, 3> & mean, const std::array<float, 3> & std) {
```
Comment (fmz) [inline src/llama-vision.cpp:253]: This might be a rootcause for some bugs ;)

**Comment by Nekotekina** [comment]
Forgive me for (probably) offtopic, but would it be possible to use CLIP model with llama.cpp to compute text embedding for the "classic" purpose of matching text with images?

File: src/llama-vision.cpp (hunk 3 above)
Comment (ngxson) [reply src/llama-vision.cpp:253]: Yes, probably missing `clip_image_f32 &`; thanks for pointing this out.

**Comment by ngxson** [comment]
> Forgive me for (probably) offtopic, but would it be possible to use CLIP model with llama.cpp to compute text embedding for the "classic" purpose of matching text with images?

I'm not 100% sure, but in theory, everything is possible with the correct model weight and the correct compute graph.

**Comment by danbev** [comment]
I took a shot at rebasing this branch and as there have been quite a few changes upstream which affects this PR and I wanted try this API out. I wanted to share the rebased [repo](https://github.com/danbev/llama.cpp/tree/vision-api-rebased) in case it would be useful/save time. 

I moved the example to a new example named [simple-vision](https://github.com/danbev/llama.cpp/tree/687d714a9a39a6ab98a5e765c5ee22522d1e7e18/examples/simple-vision#readme) as the `simple` example in upstream has been updated to remove the dependencies on common. There are probably some improvements to be made in the changes made during rebasing but it might be easier/quicker to make those changes to the rebased branch.

**Comment by sragrawal** [comment]
@ngxson are you still planning on working on this? Having this as a first step towards llama-3.2-vision would be very useful.

**Comment by ngxson** [comment]
I'm not actively working on it, but will continue soon. Ref discussion: https://github.com/ggerganov/llama.cpp/issues/10381

**Comment by danbev** [comment]
@ngxson I've been taking a look at this and tried adding some initial support for Llama 3.2 Vision Instruct. I've added a [simple-vision-mllama-example](https://github.com/danbev/llama.cpp/tree/vision-api-mllama-example/examples/simple-vision-mllama#simple-vision-mllama-example) that shows the usage also contains some more details.

The code needs more work and possibly integration with the existing vision api, but my only goal was only to get something working as a first step. If this looks like it would be worth pursuing I'd be happy to continue working on this.
